### PR TITLE
Added simulated test function validators

### LINKED ISSUE #1028: SimpleSimulator: Add a component for validating behaviour
### Description of the task

The simulator and simulated tests need ways of validating behaviour. Because this behaviour occurs over time and include several distinct "stages" this requires something more complicated than simple googletest asserts.

Because each robot may be doing something independent, there will be separate validation for each robot, as well as global validation. This will be accomplished by "validation functions" (name TBD).

`ValidationFunctions` will be lambda functions that accept a pointer to the `World` and a coroutine. The reason the validation functions will be written with coroutines is because these functions will generally be checking for "stages" in behaviour (like how Tactics and Plays are written), so coroutines make that much easier to write.

The `ValidationFunctions` may look something like this: 
```
using ValidationCoroutine = boost::coroutines2::coroutine<std::optional<std::string>>;
using ValidationFunction = std::function<void(ValidationCoroutine::push_type &, std::shared_ptr<World>)>;
```

The coroutine for the validation function will return an optional string, which will allow the function to report and errors if they are encountered. These error strings can then be propagated back to the `SimulatorBackend` or test itself to be reported by GoogleTest so failures show up nicely when the test is run.

The `SimulatorBackend` will accept a list of these `ValidationFunctions` when the simulation is started, and run each function after every simulation step. We will recommend that every robot has its own validation function, and that a "global" validation function is also provided to make more generic assertions like no robots collided.

The benefits of these per-robot functions is that we can re-use them once we write them for a given role. For example, once we write a ValidationFunction for a goalie (eg. robot never leaves the defense area, and the ball never enters the net), we can re-use it when testing other plays since all plays have a goalie. This promotes lots of code re-use and should make writing tests for new plays much easier.

Finally, since `ValidationFunctions` have a pointer to the `World` they can also make changes to the `World` state. This is useful if, for example, we wanted the World's gamestate to change once a robot got close to the ball during kickoff. With coroutines in the ValidationFunctions we can easily wait until that condition  is met and then change the gamestate before moving on to the next stage.

### Acceptance criteria

- [ ] Create a `ValidationFunction` structure using coroutines
- [ ] Make the `SimulatorBackend` take `ValidationFunctions` during simulation
- [ ] Report `ValidationFunction` failures to GoogleTest

### PR DESCRIPTION
**Pull request by MathewMacDougall**
### Description

Defined the signature of "Validation Functions" which will eventually be used to validate behavior in test with the simulator.

Created a `FunctionValidator` class to wrap and manage the coroutines and execution of a given validation function.

In general, this also happens to be a nice simple example of coroutines.

### Testing Done

Added unit tests

### Resolved Issues

resolves #1028 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Comment by MathewMacDougall** [comment]
@jared-paul @LiCody @amrm3lm please review ASAP!

**Review by Anthon1e — COMMENTED** [review]
oops this is my friends account - jared

**Review by jared-paul — COMMENTED** [review]
Looks good!

One question: how will this work for testing behaviours that have multiple states that may overlap/interfere with eachother

File: src/software/backend/simulation/validation/function_validator_test.cpp [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+    result               = function_validator.executeAndCheckForSuccess();
+    EXPECT_FALSE(result);
+
+    world->mutableBall() = Ball(Point(1, 1), Vector(0, 0), Timestamp::fromSeconds(0));
+    result               = function_validator.executeAndCheckForSuccess();
+    // Even though both conditions (x > 0 and y > 0) are met in this state, they were not
+    // met in the order the validation function wanted so the function has not succeeded /
+    // passed
+    EXPECT_FALSE(result);
+}
```
Comment (garethellis0) [inline src/software/backend/simulation/validation/function_validator_test.cpp:190]: Could we add a test where we check something in the coroutine?

**Review by garethellis0 — COMMENTED** [review]
Seems sane overall, but I definitely need to see it used in context (out-of-scope for this PR, just a note).

**Comment by MathewMacDougall** [comment]
> how will this work for testing behaviours that have multiple states that may overlap/interfere with eachother

Most behavior is independent, and per-robot. ie. Each robot generally has a sequence of actions we want to check for, and most of the time they don't need to be perfectly synchronized.

If we do need to wait for 2 or more robots to meet a condition before moving on, you can make use of an extra "global" validation function that is responsible for looking at everything and managing higher-level states if needed. For example, changing the refbox gamestate once something happens.

This will hopefully be more clear once these start getting used in my next PR.

**Review by LiCody — COMMENTED** [review]
Looks good to me, same as Gareth I'm curious how this would look like for more practical tests.

File: src/software/backend/simulation/validation/function_validator_test.cpp (hunk 1 above)
Comment (MathewMacDougall) [reply src/software/backend/simulation/validation/function_validator_test.cpp:190]: What do you mean exactly?

**Review by amrm3lm — DISMISSED** [review]
looks good and makes sense, 

A question about testing tactics in general: do we always know, given the state of world, what the chosen tactics should be/how many stages there will be? Can validation functions be made to test multiple acceptable decisions a robot may do?

**Comment by MathewMacDougall** [comment]
> A question about testing tactics in general: do we always know, given the state of world, what the chosen tactics should be/how many stages there will be? Can validation functions be made to test multiple acceptable decisions a robot may do?

Good question. Generally yes given the state of the world we know what Play will be chosen (or in the case of tests we may choose it explicitly ourselves), and therefore what tactics will run. As of right now we don't write Plays or tactics with random elements and try keep everything deterministic, so that we can reason about what should happen given the state of the world. While there are definitely multiple correct decisions a tactic can make, each one is generally best in a specific scenario and we have not yet come across a case where it is better to do something completely randomly.

(The only slightly random part is Play selection of more than 1 Play is able to run at a time, but currently this doesn't really happen. For tests we will specify the play manually to avoid this randomness)

So overall test writers should be able to know which decisions Plays and Tactics should be making given the state of the World because our logic is deterministic overall.

Hope that answers your question.

**Review by akhilveeraghanta — DISMISSED** [review]
LG:tm: :shipit:

File: src/software/backend/simulation/validation/function_validator_test.cpp (hunk 1 above)
Comment (garethellis0) [reply src/software/backend/simulation/validation/function_validator_test.cpp:190]: We would eventually like to do stuff like `EXPECT_EQUAL` in the coroutines right? Could we do something like that there.

Comment (MathewMacDougall) [reply src/software/backend/simulation/validation/function_validator_test.cpp:190]: Oh gotcha. Added a test.

**Review by garethellis0 — DISMISSED** [review]
Assuming that you manually tested that added something like `EXPECT_TRUE(false)` to the coroutine in the last test and it failed, LGTM :+1: :tada:

**Comment by MathewMacDougall** [comment]
@garethellis0 I did.

Fixed CI please re-approve.

**Comment by MathewMacDougall** [comment]
Can I get re-approval plz?

**Review by jared-paul — APPROVED** [review]
LGTM!

**Review by LiCody — APPROVED** [review]
LGTM

**Review by amrm3lm — APPROVED** [review]

**Review by garethellis0 — APPROVED** [review]

**Pull request merged**